java/arcs/core/storage/CapabilitiesResolver.kt

```diff
@@ -13,25 +13,64 @@ package arcs.core.storage
 
 import arcs.core.common.ArcId
 import arcs.core.data.Capabilities
+import arcs.core.data.Schema
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
 import arcs.core.util.TaggedLog
 
 /**
  * [CapabilitiesResolver] is a factory class that creates [StorageKey]s based on the registered
  * providers and given [Capabilities].
  */
 class CapabilitiesResolver(
-    val options: StorageKeyOptions,
+    val options: CapabilitiesResolverOptions,
     private val creators: StorageKeyCreatorsMap = getAllCreators()
 ) {
     private val log = TaggedLog { "CapabilitiesResolver" }
 
     /* Options used to construct [CapabilitiesResolver] */
-    data class StorageKeyOptions(val arcId: ArcId)
+    data class CapabilitiesResolverOptions(val arcId: ArcId)
+
+    /**
+     * Options passed to registered [StorageKey] constructors.
+     * @property arcId An identifier of an Arc requesting the [StorageKey]
+     * @property entitySchema A schema of an entities that will be stored
+     * @property unique A unique component of the [StorageKey]
+     * @property location A memory location of the [StorageKey]
+     */
+    interface StorageKeyOptions {
+        val arcId: ArcId
+        val entitySchema: Schema
+        val unique: String
+            get() = TODO("not implemented")
+        val location: String
+            get() = TODO("not implemented")
+    }
+
+    data class ContainerStorageKeyOptions(
+        override val arcId: ArcId,
+        override val entitySchema: Schema
+    ) : StorageKeyOptions {
+        override val unique: String
+            get() = ""
+        override val location: String
+            get() = arcId.toString()
+    }
+
+    data class BackingStorageKeyOptions(
+        override val arcId: ArcId,
+        override val entitySchema: Schema
+    ) : StorageKeyOptions {
+        override val unique: String
+            get() = entitySchema.name?.name.orEmpty()
+        override val location: String
+            get() = unique
+    }
 
     /* Creates and returns a [StorageKey] corresponding to the given [Capabilities]. */
     fun createStorageKey(
         capabilities: Capabilities,
-        entitySchemaHash: String? = null
+        entitySchema: Schema,
+        handleId: String
```

Comment:
Does this mean that if we have two handles capable of updating the same data, they'll be writing to different storage keys?

Reply:
Both handles will share the same storage key. In the Allocator, when fresh storage keys are created for 'create' fates,  handles get the same key, e.g.

```
recipe Foo
   blah: create
   Particle1
      foo: reads blah
   Particle2
     bar: reads blah
```

Handle 'foo', and handle 'bar' both reference the same storage key 'blah'

Reply:
I'm confused as to the semantics of `handleId` in this function then.

Reply:
in the recipe above `blah` is a recipe handle and it has a unique id.
both `foo` and `bar` handle connections have the same handle id, because they are connected to the same handle.
(note: `foo` and `bar` aren't handles, they are handle connections, there is only one handle in this recipe)

if there are 2 handles connected to the same storage key, at resolution phase (for PA it's at compile time) they'll be given the same Id and same storage key.